py: Fix compilation of persistentcode with uncommon configs

quark-zju opened the pull request:
With the following configuration:

    #define MICROPY_PERSISTENT_CODE_SAVE (1)
    #define MICROPY_PERSISTENT_CODE_LOAD (1)

The code fails to compile:

    persistentcode.c: In function 'load_raw_code':
    persistentcode.c:457:13: error: 'n_obj' undeclared
             n_obj, n_raw_code,

If MICROPY_EMIT_NATIVE disabled, there are more errors due to the use of
undefined fields in mp_raw_code_t.

This patch fixes compilation by avoiding undefined fields.

quark-zju committed:
py: Fix compilation of persistentcode with uncommon configs

With the following configuration:

    #define MICROPY_PERSISTENT_CODE_SAVE (1)
    #define MICROPY_PERSISTENT_CODE_LOAD (1)

The code fails to compile:

    persistentcode.c: In function 'load_raw_code':
    persistentcode.c:457:13: error: 'n_obj' undeclared
             n_obj, n_raw_code,

If MICROPY_EMIT_NATIVE disabled, there are more errors due to the use of
undefined fields in mp_raw_code_t.

This patch fixes compilation by avoiding undefined fields.

MICROPY_EMIT_NATIVE was changed to MICROPY_EMIT_ANY_ASM in this file to
match mp_raw_code_t definition.

dpgeorge commented on py/persistentcode.c:
```diff
@@ -408,12 +408,15 @@ STATIC mp_raw_code_t *load_raw_code(mp_reader_t *reader, qstr_window_t *qw) {
 
     mp_uint_t *const_table = NULL;
+    size_t n_obj = 0;
+    size_t n_raw_code = 0;
```
Since `n_obj` and `n_raw_code` are set before `const_table`, it makes sense to put their declarations here before it, to keep the logical order of these variables.

dpgeorge commented on py/persistentcode.c:
```diff
@@ -633,6 +636,7 @@ STATIC void save_raw_code(mp_print_t *print, mp_raw_code_t *rc, qstr_window_t *q
         // Save native code
```
This entire "else" block can be guarded by `MICROPY_EMIT_NATIVE || MICROPY_EMIT_INLINE_ASM`, instead of the two separate `#if`'s below.

quark-zju replied:
Makes sense. I'll update the reader part too.

quark-zju committed:
py: define EMIT_ANY_ASM = EMIT_NATIVE || EMIT_INLINE_ASM

There are a lot of places testing MICROPY_EMIT_NATIVE || MICROPY_EMIT_INLINE_ASM.

Define a new macro for it.

dpgeorge commented on py/mpconfig.h:
```diff
@@ -329,6 +329,9 @@
 #define MICROPY_EMIT_INLINE_ASM (MICROPY_EMIT_INLINE_THUMB || MICROPY_EMIT_INLINE_XTENSA)
 
+// Convenience definition for whether any native or inline assembler emitter is enabled
+#define MICROPY_EMIT_ANY_ASM (MICROPY_EMIT_NATIVE || MICROPY_EMIT_INLINE_ASM)
```
Yes, it's probably a good idea to make this define.  But, I'm just thinking if there's a better name for it, because "any asm" could be mistaken for meaning "any inline asm" (which is already what `MICROPY_EMIT_INLINE_ASM` is for).

Maybe `MICROPY_EMIT_MACHINE_CODE`?

dpgeorge commented:
@quark-zju I'm happy to make the final changes (rename to `MICROPY_EMIT_MACHINE_CODE`) and merge if you aren't able to.

pmp-p commented:
MICROPY_EMIT_MACHINE_CODE options seems a very good way to open doors to the new WASM inlining possibility javascript port can offer.

dpgeorge commented:
Ok, I renamed `MICROPY_EMIT_ANY_ASM` to `MICROPY_EMIT_MACHINE_CODE` and merged in b152bbddd132cf3f147a526efebedaf25eba29cd and d165a401dce66ba952b016d116b60e77b11f3e1f
